components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLLogoutHandler.java

```diff
@@ -0,0 +1,112 @@
+import org.wso2.carbon.identity.event.IdentityEventConstants.EventName;
+import org.wso2.carbon.identity.event.IdentityEventConstants.EventProperty;
+import org.wso2.carbon.identity.event.IdentityEventException;
+import org.wso2.carbon.identity.event.event.Event;
+import org.wso2.carbon.identity.event.handler.AbstractEventHandler;
+
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * SAML Logout Handler do Single Logout when session is terminated.
+ */
+public class SAMLLogoutHandler extends AbstractEventHandler {
+
+    private static Log log = LogFactory.getLog(SAMLLogoutHandler.class);
+
+    private SAMLSSOService samlSSOService = new SAMLSSOService();
+
+    @Override
+    public void handleEvent(Event event) throws IdentityEventException {
+
+        String samlssoTokenId = null;
+        String issuer = null;
+
+        if (StringUtils.equals(event.getEventName(), EventName.SESSION_TERMINATE.name())) {
+            samlssoTokenId = getSamlSSOTokenIdFromEvent(event);
+            if (StringUtils.isNotBlank(samlssoTokenId)) {
+                if (!isIDPInitiatedLogoutRequest(event)) {
+                    issuer = this.getIssuerFromContext(event);
+                }
+
+                try {
+                    samlSSOService.doSingleLogout(samlssoTokenId, issuer);
+                } catch (IdentityException e) {
+                    log.error("Error while SAML Logout Listener is doing single logout .", e);
+                }
+            } else {
+                if (log.isDebugEnabled()) {
+                    AuthenticationContext context = (AuthenticationContext) event.getEventProperties().
+                            get(EventProperty.CONTEXT);
+                    log.debug("There are no SAML participants in the session : " + context.getSessionIdentifier());
+                }
+            }
+        }
+    }
+
+    @Override
+    public String getName() {
+
+        return "SAMLLogoutHandler";
+    }
+
+    private String getSamlSSOTokenIdFromEvent(Event event) {
+
+        String samlssoTokenId = null;
+        if (event.getEventProperties().get(EventProperty.REQUEST) instanceof HttpServletRequest) {
+            HttpServletRequest request = (HttpServletRequest) event.getEventProperties().get(EventProperty.REQUEST);
+            if (request != null) {
+                Cookie cookie = FrameworkUtils.getCookie(request, SAMLSSOConstants.SAML_SSO_TOKEN_ID_COOKIE);
+                if (cookie != null) {
+                    samlssoTokenId = cookie.getValue();
+                }
+            }
+        }
+        return samlssoTokenId;
+    }
+
+    private boolean isIDPInitiatedLogoutRequest(Event event) {
+
+        boolean isIdpInitiated = true;
+        HttpServletRequest request = (HttpServletRequest) event.getEventProperties().get(EventProperty.REQUEST);
+        String slo = request.getParameter(SAMLSSOConstants.QueryParameter.SLO.toString());
```

Comment:
can we check whether the request is null since we are not doing a input argument validation at the top level